src/operations/resolve.ts

```diff
@@ -0,0 +1,159 @@
+import { Store } from '../store';
+import {
+  Data,
+  SelectionSet,
+  OperationRequest,
+  Variables,
+  Fragments,
+  DataField,
+} from '../types';
+import {
+  getSelectionSet,
+  getMainOperation,
+  normalizeVariables,
+  getFragments,
+  getName,
+  getFieldArguments,
+  getFieldAlias,
+} from '../ast';
+import { SelectionIterator } from './shared';
+import { joinKeys, keyOfField, isDataOrKey } from '../helpers';
+import warning from 'warning';
+
+interface Context {
+  store: Store;
+  variables: Variables;
+  fragments: Fragments;
+}
+
+export const resolveData = (
+  store: Store,
+  request: OperationRequest,
+  data: Data
+) => {
+  const operation = getMainOperation(request.query);
+  const ctx = {
+    store,
+    variables: normalizeVariables(operation, request.variables),
+    fragments: getFragments(request.query),
+  };
+
+  resolveDataSelection(ctx, 'Query', getSelectionSet(operation), data);
+  return data;
+};
+
+const resolveResolverResult = (
+  ctx: Context,
+  result: DataField,
+  key: string,
+  select: SelectionSet,
+  prevData: void | Data | Data[]
+) => {
+  // When we are dealing with a list we have to call this method again.
+  if (Array.isArray(result)) {
+    // @ts-ignore: Link cannot be expressed as a recursive type
+    return result.map((childResult, index) => {
```

Comment:
While we're at it we could unroll this loop here like I've done for the other map